File: Haskell/silan_paradigm.hs
step :: [[Int]] -> [[Int]]
step grid =
  [[newCell i j | j <- [0..cols-1]] | i <- [0..rows-1]]
  where
    rows = length grid
    cols = length (head grid)

    get i j =
      if i < 0 || i >= rows || j < 0 || j >= cols
      then 0
      else grid !! i !! j

    ones =
      [(i,j) | i <- [0..rows-1], j <- [0..cols-1], get i j == 1]

    infected =
      concat
        [ [(i,j), (i,j+1), (i,j-1), (i+1,j), (i-1,j)]
        | (i,j) <- ones
        ]

    newCell i j =
      if (i,j) `elem` infected then 1 else 0




applyN :: Int -> (a -> a) -> a -> a
applyN n f x 
    | n == 0 = x
    | otherwise = applyN(n - 1) f (f x) 


sqrtStep :: Double -> Double -> Double
sqrtStep a x = (x + a / x) / 2

sqrtNewton :: Int -> Double -> Double
sqrtNewton n a = applyN n (sqrtStep a) 1

File: Haskell/silan_paradigm.py
def step(grid):
    rows = len(grid)
    cols = len(grid[0])

    new_grid = []

    for i in range(rows):
        row = []
        for j in range(cols):
            me = grid[i][j]
            up = grid[i-1][j] if i > 0 else 0
            down = grid[i+1][j] if i < rows-1 else 0
            left = grid[i][j-1] if j > 0 else 0
            right = grid[i][j+1] if j < cols-1 else 0

            value = max(me, up, down, left, right)
            row.append(value)

        new_grid.append(row)

    return new_grid


grid = [
    [0,0,0,0,0],
    [0,1,0,0,0],
    [0,0,0,0,0],
    [0,0,0,1,0],
    [0,0,0,0,0]
]

print(step(grid))

def sqrt_newton(a, n):
    x = 1

    for i in range(n):
        x = (x + a/x) / 2

    return x

print(sqrt_newton(9,5))

